Add endpoint for GSSAPI Authentication

antoniotorresm opened the pull request:
Add the `login_kerberos` endpoint for handling GSSAPI authentications. mod_auth_gssapi and gssproxy are included as dependencies. Additional steps are included to the IPA domain addition, such as the addition of the HTTP service and keytab retrieval. Additionally, `login_password` endpoint is provided as well, which requests a ticket using the user and password passed with the client request.

abbra commented on src/ipa-tuura/domains/utils.py:
```diff
@@ -212,8 +212,10 @@ def deploy_ipa_service(domain):
     hostname = socket.gethostname()
     realm = domain["name"].upper()
     ipatuura_principal = "ipatuura/%s@%s" % (hostname, realm)
+    http_principal = "HTTP/%s@%s" % (hostname, realm)
```
Here is the problem. We should not use bridge's hostname here. Instead our keytab should contain a key which will correspond to the keycloak's host. I think you should make a separate confgiuration entry for this.

The idea is that this whole setup will be:

client browser -> Keycloak service -> IPA bridge

and GSSAPI authentication would happen between the client browser and Keycloak service. From that service a headers for that exchange would be forwarded to the bridge and then to the login end-point we have created. Here they'll be processed.

From GSSAPI point of view the client would ask for a service ticket to the Keycloak service host. It would be `HTTP/keycloak-hostname`, not `HTTP/ipatuura-hostname`. Since we'd have the keytab for `HTTP/keycloak-hostname` service key, we will be able to decode it.

antoniotorresm replied:
What do you think about receiving the keycloak hostname as part of the request and using that for creating the `HTTP/keycloak-hostname` service and requesting the keytab during domain creation? I have already added the automation for `setspn.exe`, if we have the keycloak hostname we could automate keytab retrieval as well.

antoniotorresm commented:
Added a new field to the domain model, `keycloak_hostname` that will take the hostname for the Keycloak host. I have also added automation for SPN creation and keytab retrieval on AD hosts.

antoniotorresm commented:
/ok-to-test

antoniotorresm commented:
/retest

f-trivino reviewed (approved):
LGTM!
I have tested the new end-points for IPA and AD use cases using the same curl calls:

`curl -k -v -X POST --negotiate -u : https://bridge.ipa.test/bridge/login_kerberos/`

` curl -k -v -X POST https://bridge.ipa.test/bridge/login_password/ --data 'user=ipauser&password=Secret123'`

I think the PR covers the initial feature for IPA and AD, and it can be easily extended later on to cover RHDS/LDAPwithKerberos. 

Ref testing, we are enabling RedHatTrustedPipeline and the plan is to run extensive e2e tests there.

abbra commented on src/ipa-tuura/domains/utils.py:
```diff
@@ -291,6 +329,16 @@ def deploy_ipa_service(domain):
+    proc = subprocess.run(args, capture_output=True, text=True)
+    if proc.returncode != 0:
+        raise Exception("Error getkeytab:\n{}".format(proc.stderr))
+
```
For this whole block of `ipa-getkeytab` use -- check that `http_principal` is different from `http_bridge_principal` and only then run the second one. Otherwise you are invalidating the first one.

abbra commented on src/ipa-tuura/domains/utils.py:
```diff
@@ -374,6 +422,39 @@ def join_ad_realm(domain):
+        f"setspn -S HTTP/{kc_hostname} ipatuura;"
+        f"setspn -S HTTP/{bridge_hostname} ipatuura;"
+        "$kvno = Get-ADuser ipatuura -property msDS-KeyVersionNumber | select -expand msDS-KeyVersionNumber;"
+        f"ktpass -out /httpd.keytab -mapUser ipatuura@{ad_realm} -pass {ad_passwd} -mapOp set +DumpSalt -crypto AES256-SHA1 -ptype KRB5_NT_PRINCIPAL -princ HTTP/{kc_hostname}@{ad_realm} -kvno $kvno;"
```
You might want to allow encryption type list be configurable. There are plenty of environments where people are still using RC4 (sorry) or AES128-SHA1. Also, with Windows Server 2025 there will be support for RFC8009 encrytpion types (SHA2-based) so it is best to remove the hard-coded encryption lines here and instead add a configuration with sensible defaults.

antoniotorresm replied:
Created ticket #100 to address this.

abbra commented on src/ipa-tuura/domains/utils.py:
```diff
@@ -374,6 +422,39 @@ def join_ad_realm(domain):
+            "-p",
+            ad_passwd,
+            "scp",
+            f"Administrator@{ad_server}:C:/httpd.keytab",
```
Same here with `Administrator` -- the account used for AD (and IPA admin, for that reason) must be configurable.

abbra reviewed (changes requested):
I added few in-line comments.

antoniotorresm commented:
Added `Remote-User` response header, needed for Keycloak plugin.
